File: src_fortran/sub/spreadb.f
c
c
c
c
c
c
      subroutine SPREADb(R,m,spread) ! Urs Kradolfer
c     BACKUS-GILBERT SPREAD FUNCTION
c     measures the goodness of the resolution spread, based on the
c     L2 norm of the WEIGHTED difference between the resolution matrix and the
c     identity matrix [ Backus-Gilbert spread function ]
c     R = I   <==>  spread(R) = 0
c
      implicit none
      integer m
      real r(m,m)
      real spread
      integer j,i
c
      spread=0.0
      do i=1,m
         do j=1,m
            spread=spread+(i-j)*(i-j)*r(i,j)*r(i,j)
         enddo
      enddo
      return
      end ! of subr. spreadb

File: src/sub/spreadb.c
#include <stddef.h>

void spreadb(const double *r, int m, double *spread) {
    int i;
    int j;

    if (r == NULL || m <= 0 || spread == NULL) {
        return;
    }

    *spread = 0.0;
    for (i = 0; i < m; ++i) {
        for (j = 0; j < m; ++j) {
            double diff = (double)(i - j);
            double val = r[i * m + j];
            *spread += diff * diff * val * val;
        }
    }
}
